service/src/test/java/bio/terra/workspace/app/configuration/external/controller/FolderApiControllerTest.java

```diff
@@ -0,0 +1,310 @@
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.util.Optional;
+import java.util.UUID;
+import javax.annotation.Nullable;
+import org.apache.http.HttpStatus;
+import org.broadinstitute.dsde.workbench.client.sam.model.UserStatusInfo;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.http.MediaType;
+import org.springframework.test.web.servlet.MockMvc;
+
+public class FolderApiControllerTest extends BaseUnitTest {
+  private static final AuthenticatedUserRequest USER_REQUEST =
+      new AuthenticatedUserRequest(
+          "fake@email.com", "subjectId123456", Optional.of("ThisIsNotARealBearerToken"));
+
+  @Autowired MockMvc mockMvc;
+  @Autowired ObjectMapper objectMapper;
+
+  @MockBean SamService mockSamService;
+
+  @BeforeEach
+  public void setUp() throws InterruptedException {
+    // Needed for workspace creation as logging is triggered when a workspace is created in
+    // `WorkspaceActivityLogHook` where we extract the user request information and log it to
+    // change agents.
+    when(mockSamService.getUserStatusInfo(any()))
+        .thenReturn(
+            new UserStatusInfo()
+                .userEmail(USER_REQUEST.getEmail())
+                .userSubjectId(USER_REQUEST.getSubjectId()));
+  }
+
+  @Test
+  public void createFolder_topLevelFolderCreated() throws Exception {
+    UUID workspaceId = createDefaultWorkspace(mockMvc, objectMapper).getId();
+    var displayName = "foo";
+    var description = String.format("This is folder %s", displayName);
+
+    ApiFolderDescription folderDescription =
+        createFolder(workspaceId, displayName, description, null);
+
+    assertEquals(displayName, folderDescription.getDisplayName());
+    assertEquals(description, folderDescription.getDescription());
+    assertNotNull(folderDescription.getId());
+    assertNull(folderDescription.getParentFolderId());
+  }
+
+  @Test
+  public void createFolder_parentFolderDoesNotExist_throws404() throws Exception {
+    UUID workspaceId = createDefaultWorkspace(mockMvc, objectMapper).getId();
+    var displayName = "foo";
+    var description = String.format("This is folder %s", displayName);
+
+    mockMvc
+        .perform(
+            addJsonContentType(
+                addAuth(
+                    post(String.format(FOLDERS_V1_PATH_FORMAT, workspaceId))
+                        .content(
+                            objectMapper.writeValueAsString(
+                                createFolderRequestBody(
+                                    displayName, description, UUID.randomUUID()))),
+                    USER_REQUEST)))
+        .andExpect(status().is(HttpStatus.SC_NOT_FOUND));
+  }
+
+  @Test
+  public void createFolder_duplicateFolder_throws400() throws Exception {
```

Comment:
Can you also test that duplicate names are allowed, in different folders? This could be an example where it's useful to test more than one thing in a test method. It's easier to read than X different methods; X different methods would have overlapping setup.

```
  @Test
  public void createFolder_subFolder() throws Exception {
    UUID workspaceId = createDefaultWorkspace(mockMvc, objectMapper).getId();
    var displayName = "foo";

    createFolder(parentFolder, displayName);

    // Can't create a second sub-folder called displayName
    createFolderExpectingError(parentFolder, displayName, HttpStatus.SC_BAD_REQUEST);

    // Can create a sub-folder called displayName in a different parent folder
    createFolder(parentFolder2, displayName);
  }
```